File: dayhye/Programmers/연습문제/lv2/오픈채팅방.py
def solution(record):
    answer = []
    sts = []
    states = {}

    for rec in record:
        temp = list(rec.split())

        if temp[0] == 'Enter':
            states[temp[1]] = temp[2]
            sts.append([temp[1], "님이 들어왔습니다."])
        elif temp[0] == 'Leave':
            sts.append([temp[1], "님이 나갔습니다."])
        else:
            states[temp[1]] = temp[2]

    for id, state in sts:
        total = states[id] + state

        answer.append(total)

    return answer

File: eunnyeong/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String, String> h = new HashMap<>(); // id, 이름
        
        List<Pair> list = new ArrayList<>(); //id, 입장여부
        for (int i = 0; i < record.length; i++) {
            StringTokenizer st = new StringTokenizer(record[i]);
            String a = st.nextToken();
            String b = st.nextToken();
            String c = "";
            if (!a.equals("Leave")) {
                c = st.nextToken();
                h.put(b, c);
            }
             if (!a.equals("Change"))
                list.add(new Pair(b, a));
        }
        
        String[] answer = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            Pair p = list.get(i);
            String s = h.get(p.x) + "님이 ";
            
            if (p.y.equals("Enter"))
                s += "들어왔습니다.";
            else 
                s += "나갔습니다.";
                
            answer[i] = s;
        }
        
        return answer;
    }
}

class Pair {
    String x, y;
    
    public Pair (String x, String y) {
        this.x = x;
        this.y = y;
    }
}
